hathor/transaction/base_transaction.py

```diff
@@ -60,26 +61,6 @@
 _ONE_BYTE = 0xFF
 
 
-def sum_weights(w1: float, w2: float) -> float:
```

Comment:
I think I won't be able to remove the `sum_weights()` in this PR because it is still used in 3 places:

- `metrics.py`: Seems easy to remove it.
- `difficulty.py`: This is used by merged mining and maybe Jan can help me understand the consequences of changing it. Anyway, it seems feasible.
- `stratum/straum.py`: Seems easy but we need to better assess its impact. I think it's used in the hash rate estimator only. And, to be honest, I don't know if we still use this code.
- `daa.py`: I think we should not change this one, as it might affect the minimum weight of the blocks and invalidate blocks that are on the mainnet. To confirm the consequence, we would need to check it against the mainnet but it seems not worth the risk.

My impression is that, in the end, we will remove the `sum_weights()` from (1), (2), and (3) and then we will move its code to the `daa.py` and continue using it there.

What do you think?

Reply:
In `difficulty` it is only used to implement a `logsum` for function reuse. The following patch ports the exact same math:

```
diff --git a/hathor/difficulty.py b/hathor/difficulty.py
index 4a8467ef..fa80c124 100644
--- a/hathor/difficulty.py
+++ b/hathor/difficulty.py
@@ -787,8 +787,8 @@ class Weight(float):
 
         Currently is just a proxy to `hathor.transaction.sum_weights`.
         """
-        from hathor.transaction import sum_weights
-        return Weight(sum_weights(self, other))
+        a, b = self, other if self > other else other, self
+        return Weight(a + log2(1 + 2 ** (b - a)))
 
     def to_u256(self) -> U256:
         """ Convert to U256

```

Since `hathor.difficulty` is intended to abstract that kind of values, it can be used in places like `hathor.metrics`, `hathor.stratum.stratum` and `hathor.daa`. That should preserve the exact same operations as before. I can see how `hathor.daa` could maybe be refactored but I agree it's not worth to touch it now.

Reply:
I've decided to keep `sum_weights` so we don't have to refactor `hathor.metrics`, `hathor.stratum.stratum`, etc.